### PR TITLE
Add `apply_suggested_lr` to LRFinder

### PR DESCRIPTION
**Pull request by KickItLikeShika**
Added `apply_suggested_lr` to LRFinder, added tests, updated tests to use MNIST, and updated the example.
Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

File: ignite/contrib/handlers/lr_finder.py [hunk 1]
```diff
@@ -310,6 +310,34 @@ def lr_suggestion(self) -> Any:
...
+            optimizer: the optimizer to apply the suggested learning rate(s) on.
+
+        Returns:
+                The optimizer with the new learning rate(s).
+        """
+        sug_lr = self.lr_suggestion()
+        if isinstance(sug_lr, list):
+            if len(sug_lr) != len(optimizer.param_groups):
+                raise RuntimeError(
+                    "Applying the suggested learning rate can't be done, the optimizers are not the same"
```
Comment (vfdev-5) [inline ignite/contrib/handlers/lr_finder.py:330]:
```suggestion
                    "The number of parameter groups does not match between "
                    "given optimizer and the one used for estimating the "
                    f"learning rate: {len(sug_lr)} vs {len(optimizer.param_groups)}"
```

File: ignite/contrib/handlers/lr_finder.py [hunk 2]
```diff
@@ -310,6 +310,34 @@ def lr_suggestion(self) -> Any:
...
+                )
+            i = 0
+            for lr in sug_lr:
+                optimizer.param_groups[i]["lr"] = lr
+                i += 1
```
Comment (vfdev-5) [inline ignite/contrib/handlers/lr_finder.py:335]:
```suggestion
            for i, lr in enumerate(sug_lr):
                optimizer.param_groups[i]["lr"] = lr
```

File: ignite/contrib/handlers/lr_finder.py [hunk 3]
```diff
@@ -310,6 +310,34 @@ def lr_suggestion(self) -> Any:
         min_grad_idx = grads.argmin() + 1
         return self._history["lr"][int(min_grad_idx)]
 
+    def apply_suggested_lr(self, optimizer: Optimizer) -> Optimizer:
+        """
+        Applying the suggested learning rate(s) on the given optimizer.
```
Comment (vfdev-5) [inline ignite/contrib/handlers/lr_finder.py:315]: Technically, it is inplace method. Let's return nothing to enphasize that.

File: examples/notebooks/FastaiLRFinder_MNIST.ipynb [hunk 4]
```diff
@@ -238,7 +238,7 @@
     "id": "NcT19wqkmvd6"
```
Comment (vfdev-5) [inline examples/notebooks/FastaiLRFinder_MNIST.ipynb:238]: Let's update this notebook as we discussed previously:
- Let's split this cell into multiple ones:
```
trainer = create_supervised_trainer(model, optimizer, criterion, device=device)
ProgressBar(persist=True).attach(trainer, output_transform=lambda x: {"batch loss": x})

lr_finder = FastaiLRFinder()
to_save={'model': model, 'optimizer': optimizer}
with lr_finder.attach(trainer, to_save, diverge_th=1.5) as trainer_with_lr_finder:
    trainer_with_lr_finder.run(trainloader)
    
trainer.run(trainloader, max_epochs=10)

evaluator = create_supervised_evaluator(model, metrics={"acc": Accuracy(), "loss": Loss(nn.NLLLoss())}, device=device)
evaluator.run(testloader)

print(evaluator.state.metrics)
```
->
```python
# CELL 1
trainer = create_supervised_trainer(model, optimizer, criterion, device=device)
ProgressBar(persist=True).attach(trainer, output_transform=lambda x: {"batch loss": x})

lr_finder = FastaiLRFinder()
to_save={'model': model, 'optimizer': optimizer}
with lr_finder.attach(trainer, to_save, diverge_th=1.5) as trainer_with_lr_finder:
    trainer_with_lr_finder.run(trainloader)
    
# CELL 2 - Text
Let's plot the found optimal learning rate.

# CELL 3
from matplotlib import pyplot as plt
ax = lr_finder.plot()
plt.axes(ax)
plt.show()

print("Suggested LR", lr_finder.lr_suggestion())

# CELL 4 - Text
Let's compare training results: a) using a fixed learning rate (`lr=3e-4`) and b) with applying optimal learning rate.

# CELL 5
trainer.run(trainloader, max_epochs=10)

evaluator = create_supervised_evaluator(model, metrics={"acc": Accuracy(), "loss": Loss(nn.NLLLoss())}, device=device)
evaluator.run(testloader)

print(evaluator.state.metrics)
```

File: examples/notebooks/FastaiLRFinder_MNIST.ipynb [hunk 5]
```diff
@@ -251,9 +251,13 @@
    },
    "outputs": [],
    "source": [
-    "optimizer.param_groups[0]['lr'] = lr_finder.lr_suggestion()\n",
+    "new_optimizer = lr_finder.apply_suggested_lr(optimizer)\n",
+    "\n",
+    "new_trainer = create_supervised_trainer(model, new_optimizer, criterion, device=device)\n",
```
Comment (vfdev-5) [inline examples/notebooks/FastaiLRFinder_MNIST.ipynb:256]: You have to use new model or restore weights from scratch.

File: tests/ignite/contrib/handlers/test_lr_finder.py [hunk 6]
```diff
@@ -337,21 +337,41 @@ def test_lr_suggestion(lr_finder, to_save, dummy_engine, dataloader):
...
 ):
 
-    with lr_finder.attach(
-        dummy_engine_single_param_groups, to_save_single_param_groups, end_lr=20, smooth_f=0.04
-    ) as trainer_with_finder:
-        trainer_with_finder.run(dataloader_plot)
+    with lr_finder.attach(dummy_engine_mnist, mnist_to_save) as trainer_with_finder:
+        trainer_with_finder.run(mnist_dataloader)
 
-    # Avoid RuntimeError
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_lr_finder.py:349]: Can you also update `test_lr_suggestion` and remove : 
```python
    # Avoid RuntimeError
    if torch.tensor(lr_finder._history["loss"]).argmin() < 2:
        lr_finder._history["loss"].insert(0, 10)
        lr_finder._history["loss"].insert(0, 100)
        lr_finder._history["lr"].insert(0, 10)
        lr_finder._history["lr"].insert(0, 100)

    assert 1e-4 <= lr_finder.lr_suggestion() <= 10
```
from there ?

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @KickItLikeShika 
Few comments to fix

File: tests/ignite/contrib/handlers/test_lr_finder.py (hunk 6 above)
Comment (KickItLikeShika) [reply tests/ignite/contrib/handlers/test_lr_finder.py:349]: I thought about that, but If we did that we will have to MNIST dataset, and if we did that, there will be no difference between `test_lr_suggestion` and `test_mnist_lr_suggestion`, so we can leave `test_lr_suggestion` as it is or we can delete it and rename `test_mnist_lr_suggestion` to `test_lr_suggestion`, and consider it the learning rate test function, what do you think?

Comment (vfdev-5) [reply tests/ignite/contrib/handlers/test_lr_finder.py:349]: Let's do the following: let's manually create loss/lr history without calling
```python
    with lr_finder.attach(dummy_engine, to_save) as trainer_with_finder:
        trainer_with_finder.run(dataloader)
```
and check lr_suggestion on this synthetic data.
Another comment:
```python
assert 1e-4 <= lr_finder.lr_suggestion() <= 10
```
this assertion should be stricter, internal [1e-4, 10] is too large.

**Comment by KickItLikeShika** [comment]
That's strange! these failures didn't happen locally!

**Comment by KickItLikeShika** [comment]
It seems to be working fine now! Let's now rename `test_mnist_lr_suggestion` to `test_lr_suggestion_mnist`. And that's will help us double check everything

File: examples/notebooks/FastaiLRFinder_MNIST.ipynb [hunk 7]
```diff
@@ -238,11 +255,36 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# We can reset the parameters of our model\n",
+    "# or we can just create new instance for the model\n",
+    "new_model = Net()\n",
+    "new_model.to(device)\n",
```
Comment (vfdev-5) [inline examples/notebooks/FastaiLRFinder_MNIST.ipynb:282]: If you use new model, you have to set parameters to the optimizer as it has old model's params right now

File: tests/ignite/contrib/handlers/test_lr_finder.py [hunk 8]
```diff
@@ -323,35 +317,105 @@ def test_lr_suggestion_unexpected_curve(lr_finder, to_save, dummy_engine, datalo
...
+    # 2 param_groups
+    lr_finder._history["lr"] = torch.tensor(np.linspace(0.01, 10, num=100)).reshape(50, 2)
+
+    # lr_finder.lr_suggestion() is supposed to return a list of values,
+    # but as we assign loss and lr to tensors, instead of lists, it will return tensors
+    suggested_lrs = lr_finder.lr_suggestion()
+
+    # Ignoring the increasing part of the curve in the assertion.
+    assert 0.1 <= suggested_lrs[0].item() <= 6
+    assert 0.1 <= suggested_lrs[1].item() <= 6
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_lr_finder.py:348]: Can we make these assertions more precise: `pytest.approx(suggested_lrs[0]) == 0.xyz`

File: tests/ignite/contrib/handlers/test_lr_finder.py [hunk 9]
```diff
@@ -323,35 +317,105 @@ def test_lr_suggestion_unexpected_curve(lr_finder, to_save, dummy_engine, datalo
...
-        lr_finder._history["lr"].insert(0, 10)
-        lr_finder._history["lr"].insert(0, 100)
+    with lr_finder.attach(dummy_engine_mnist, mnist_to_save) as trainer_with_finder:
+
+        with trainer_with_finder.add_event_handler(
+            Events.ITERATION_COMPLETED(once=max_iters), lambda _: trainer_with_finder.terminate()
+        ):
+            trainer_with_finder.run(mnist_dataloader)
 
     assert 1e-4 <= lr_finder.lr_suggestion() <= 10
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/test_lr_finder.py:362]: Same here, let's reduce the interval to a smaller one

**Comment by vfdev-5** [comment]
@KickItLikeShika I'm running the notebook on the colab and I think we do not need to call `plt.axes(ax)` : 

```python
from matplotlib import pyplot as plt
ax = lr_finder.plot()
plt.axes(ax)
plt.show()

print("Suggested LR", lr_finder.lr_suggestion())
```

File: examples/notebooks/FastaiLRFinder_MNIST.ipynb [hunk 10]
```diff
@@ -251,11 +308,10 @@
    },
    "outputs": [],
    "source": [
-    "optimizer.param_groups[0]['lr'] = lr_finder.lr_suggestion()\n",
+    "new_trainer.run(trainloader, max_epochs=10)\n",
```
Comment (vfdev-5) [inline examples/notebooks/FastaiLRFinder_MNIST.ipynb:311]: I think it is better to reduce `max_epochs` to 5 epochs everywhere. In case of suggested LR it is rather large and on during the last 5 epochs batch loss jumps too much as LR is rather high.

File: examples/notebooks/FastaiLRFinder_MNIST.ipynb [hunk 11]
```diff
@@ -218,18 +218,53 @@
...
   },
   {
-   "cell_type": "code",
-   "execution_count": null,
+   "cell_type": "markdown",
    "metadata": {
     "id": "EHHNoOi8mvd6"
    },
+   "source": [
+    "Let's compare training results: \n",
```
Comment (vfdev-5) [inline examples/notebooks/FastaiLRFinder_MNIST.ipynb:232]: I think to convey a better message on how to use the LR finder we have to change the order of below runs.
The idea is to show how we can use LR finder in user code without re-creating model and optimizer again.
So, let's continue the training with suggested LR and create new model / optimizer with initial params.

By the way, the line "Note, that model and optimizer were restored to their initial states when FastaiLRFinder finished." should be put here and not below.

Comment (KickItLikeShika) [reply examples/notebooks/FastaiLRFinder_MNIST.ipynb:232]: > So, let's continue the training with suggested LR and create new model / optimizer with initial params.

Can you please explain this more?

Comment (vfdev-5) [reply examples/notebooks/FastaiLRFinder_MNIST.ipynb:232]: Think about how would you use this feature in your project:
```python
model = ...
optimizer = ...
trainer = ...

# do lr finder trick
# ...

lr_finder.apply_suggested_lr(optimizer)

trainer.run(...)
```

We do not want to create another model etc and this should be shown here as an example.

Comment (KickItLikeShika) [reply examples/notebooks/FastaiLRFinder_MNIST.ipynb:232]: I tried this approach already, but I found nothing on how to reset the parameters I found this discussion https://discuss.pytorch.org/t/reset-optimizer-stats/78516/5?u=kickitlikeshika , https://github.com/pytorch/pytorch/issues/37410

Comment (vfdev-5) [reply examples/notebooks/FastaiLRFinder_MNIST.ipynb:232]: Why would you like to reset parameters in this case ?

Comment (KickItLikeShika) [reply examples/notebooks/FastaiLRFinder_MNIST.ipynb:232]: Should we continue training with same old parameters? that will be easier to implement, but I though we needed to show the user that a new model/optimizer with found LR will perform way better, that's why I wanted to start from scratch again..

Comment (vfdev-5) [reply examples/notebooks/FastaiLRFinder_MNIST.ipynb:232]: Let me explain again my point. The notebook is nice now and almost OK, but it does not show the user how to use the feature in their project as we are creating new models/optim etc. In the doc I suggest to write it like that : 
```python
model = ...
optimizer = ...

init_model_state = model.state_dict()
init_opt_state = optimizer.state_dict()

# ----

trainer = create_supervised_trainer(model, optimizer, criterion, device=device)
ProgressBar(persist=True).attach(trainer, output_transform=lambda x: {"batch loss": x})

lr_finder = FastaiLRFinder()
to_save={'model': model, 'optimizer': optimizer}
with lr_finder.attach(trainer, to_save, diverge_th=1.5) as trainer_with_lr_finder:
    trainer_with_lr_finder.run(trainloader)

# ----

from matplotlib import pyplot as plt
ax = lr_finder.plot()
plt.show()

print("Suggested LR", lr_finder.lr_suggestion())

# ----

Let's compare training results:

    a) with applying optimal learning rate.
    b) using a fixed learning rate (lr=3e-4)

# ----

lr_finder.apply_suggested_lr(optimizer)
print(optimizer.param_groups[0]['lr'])

# ----

trainer.run(trainloader, max_epochs=10)

evaluator = create_supervised_evaluator(model, metrics={"acc": Accuracy(), "loss": Loss(nn.NLLLoss())}, device=device)
evaluator.run(testloader)

print(evaluator.state.metrics)

# ----

# Reinit model / optim
model.load_state_dict(init_model_state)
optimizer.load_state_dict(init_opt_state)

# ----

trainer.run(trainloader, max_epochs=10)

evaluator = create_supervised_evaluator(model, metrics={"acc": Accuracy(), "loss": Loss(nn.NLLLoss())}, device=device)
evaluator.run(testloader)

print(evaluator.state.metrics)

```

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks for the PR @KickItLikeShika !
